### PR TITLE
First version of Service which reads Firehose Stream

### PR DESCRIPTION
**Pull request by pkajaba**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/python3
+
+import os
+from setuptools import setup, find_packages
+
+
+def get_requirements():
+    requirements_txt = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'requirements.txt')
+    with open(requirements_txt) as fd:
+        return fd.read().splitlines()
```
Comment (fridex) [inline setup.py]: This will return also commented out lines (probably ok):

```python
def get_requirements():
    requirements_txt = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'requirements.txt')
    with open(requirements_txt) as fd:
        lines = fd.read().splitlines()

    return list(line for line in lines if not line.startswith('#'))
```

File: f8a_firehose_fetcher/firehose_fetcher.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            'ecosystem': ecosystem,
+            'name': name,
+            'version': version
+        }
+
+        self.log.debug("Scheduling Selinon flow '%s' with node_args: '%s'", 'bayesianFlow', node_args)
+        return run_flow('bayesianFlow', node_args)
+
+    def run(self):
+        def with_requests(url):
```
Comment (fridex) [inline f8a_firehose_fetcher/firehose_fetcher.py]: Is there any intention by introducing this nested function?

File: f8a_firehose_fetcher/firehose_fetcher.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            """Get a streaming response for the given event feed using requests."""
+            import requests
+            return requests.get(url, stream=True)
+
+        self.log.debug("Initializing Server Side Events client")
+        response = with_requests(STREAM_URL)
```
Comment (fridex) [inline f8a_firehose_fetcher/firehose_fetcher.py]: Missing response check with subsequent error log.

File: setup.py (hunk 1 above)
Comment (pkajaba) [inline setup.py]: Thanks, this will produce nicer ouput!

File: f8a_firehose_fetcher/firehose_fetcher.py (hunk 2 above)
Comment (pkajaba) [inline f8a_firehose_fetcher/firehose_fetcher.py]: no, it's taken from example :-) I am about to simplify it.

File: f8a_firehose_fetcher/firehose_fetcher.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        client = sseclient.SSEClient(response)
+        for event in client.events():
+            data = json.loads(event.data)
+            self.log.debug("Received event for package '%s' from ecosystem '%s'", data.get('name'),
+                           data.get('platform').lower())
+            # For current set of ecosystems (nuget, pypi, maven and npm) calling lower is sufficient
+            self.analyses_selinon_flow(data.get('name'), data.get('version'), data.get('platform').lower())
```
Comment (jpopelka) [inline f8a_firehose_fetcher/firehose_fetcher.py]: Libraries.io supports over 30 ecosystems (platforms), while fabric8-analytics only 4. Forwarding events from all platforms would IMHO create a lot of noise. What about reducing the set of platforms to the known ones ?

Comment (pkajaba) [inline f8a_firehose_fetcher/firehose_fetcher.py]: That's a very important catch. Thanks, I will change it.

File: f8a_firehose_fetcher/firehose_fetcher.py (hunk 3 above)
Comment (pkajaba) [inline f8a_firehose_fetcher/firehose_fetcher.py]: Should be resolved in another commit

File: f8a_firehose_fetcher/firehose_fetcher.py (hunk 4 above)
Comment (pkajaba) [inline f8a_firehose_fetcher/firehose_fetcher.py]: Should be fixed.

File: f8a_firehose_fetcher/firehose_fetcher.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            self.log.debug("Unable to retrieve connection with '%s'", STREAM_URL)
+            response.raise_for_status()
+
+        client = sseclient.SSEClient(response)
+        for event in client.events():
+            data = json.loads(event.data)
+            self.log.debug("Received event for package '%s' from ecosystem '%s'", data.get('name'),
+                           data.get('platform').lower())
+
+            ecosystem = data.get('platform').lower()
```
Comment (jpopelka) [inline f8a_firehose_fetcher/firehose_fetcher.py]: If there's no `'platform'` in `data`, you get quite cryptic `AttributeError: 'NoneType' object has no attribute 'lower'`, because the `get()` returns `None` in that case.

You should either use `ecosystem = data.get('platform', '').lower()` in which case you get `ecosystem = ''`

Or `ecosystem =  data['platform'].lower()` in which case you get better error message: `KeyError: 'platform'`.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env groovy
+@Library('github.com/msrb/cicd-pipeline-helpers')
+
+def commitId
+node('docker') {
+
+    def image = docker.image('bayesian/firehose-fetcher')
```
Comment (jpopelka) [inline Jenkinsfile]: I'd rename `bayesian/firehose-fetcher` -> `fabric8-analytics/firehose-fetcher`

here and in `openshift/template.yml`

Comment (pkajaba) [inline Jenkinsfile]: all images are prefixed with bayesian name. I can rename it, but it will be inconsistent.

Comment (jpopelka) [inline Jenkinsfile]: OK, I was looking at Michal's https://github.com/fabric8-analytics/fabric8-analytics-license-analysis/pull/5 where he had done the same renaming. @msrb WDYT

File: openshift/template.yml [hunk 7]
```diff
@@ -0,0 +1,139 @@
+apiVersion: v1
+kind: Template
+labels:
+  template: firehose-fetcher
```
Comment (msrb) [inline openshift/template.yml]: Please prefix these names with `f8a-` (see for example https://github.com/fabric8-analytics/fabric8-analytics-license-analysis/blob/master/openshift/template.yaml).

File: openshift/template.yml [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            valueFrom:
+              secretKeyRef:
+                name: aws
+                key: sqs-access-key-id
+          - name: AWS_SQS_SECRET_ACCESS_KEY
+            valueFrom:
+              secretKeyRef:
+                name: aws
+                key: sqs-secret-access-key
+          - name: AWS_S3_ACCESS_KEY_ID
```
Comment (msrb) [inline openshift/template.yml]: Do we need S3 secrets for this service?

File: openshift/template.yml [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            valueFrom:
+              secretKeyRef:
+               name: aws
+               key: s3-access-key-id
+          - name: AWS_S3_SECRET_ACCESS_KEY
+            valueFrom:
+              secretKeyRef:
+                name: aws
+                key: s3-secret-access-key
+          - name: BAYESIAN_SYNC_S3
```
Comment (msrb) [inline openshift/template.yml]: Nope, don't sync :)

File: openshift/template.yml [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            valueFrom:
+              secretKeyRef:
+                name: aws
+                key: sync-s3
+          # this won't get used if running with SQS
+          - name: RABBITMQ_SERVICE_SERVICE_HOST
```
Comment (msrb) [inline openshift/template.yml]: Not needed(?)

Comment (msrb) [inline openshift/template.yml]: I mean all of these variables...

File: openshift/template.yml [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                key: password
+          image: "${DOCKER_REGISTRY}/${DOCKER_IMAGE}:${IMAGE_TAG}"
+          name: firehose-fetcher
+          livenessProbe:
+            exec:
+              command:
+              - cli
+              - "--liveness"
+            initialDelaySeconds: 15
+            timeoutSeconds: 1
```
Comment (msrb) [inline openshift/template.yml]: This timeout is too lean in my opinion.

File: openshift/template.yml [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+          readinessProbe:
+            exec:
+              command:
+              - cli
+              - "--liveness"
+            initialDelaySeconds: 15
+            timeoutSeconds: 1
```
Comment (msrb) [inline openshift/template.yml]: Same here.

File: openshift/template.yml [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+          resources:
+            requests:
+              memory: "256Mi"
+            limits:
+              memory: "1024Mi"
```
Comment (msrb) [inline openshift/template.yml]: How was these limits worked out? We also want to add CPU limits, just to play it safe.

File: openshift/template.yml [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+          service: firehose-fetcher
+      spec:
+        containers:
+        - env:
+          - name: DEPLOYMENT_PREFIX
+            valueFrom:
+              configMapKeyRef:
+                name: bayesian-config
+                key: deployment-prefix
+          - name: WORKER_ADMINISTRATION_REGION
```
Comment (msrb) [inline openshift/template.yml]: I assume these variables are needed by dispacher (?)

File: Jenkinsfile [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                image.push(commitId)
+            }
+        }
+    }
+}
+
+if (env.BRANCH_NAME == 'master') {
+    node('oc') {
+
+        def dc = 'firehose-fetcher'
```
Comment (msrb) [inline Jenkinsfile]: Please add `f8a-` prefix, see my comment in the template.

File: openshift/template.yml [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    name: firehose-fetcher
+  spec:
+    selector:
+      service: firehose-fetcher
+
+parameters:
+- description: URL where libraries.io Firehose stream is running
+  displayName: Firehose Stream URL
+  required: true
+  name: FIREHOSE_STREAM_URL
```
Comment (msrb) [inline openshift/template.yml]: Please add also `ENABLE_SCHEDULING`.

File: setup.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+def get_requirements():
+    requirements_txt = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'requirements.txt')
+    with open(requirements_txt) as fd:
+        lines = fd.read().splitlines()
+
+    return list(line for line in lines if not line.startswith('#'))
+
+setup(
+    name='fabric8_analytics_firehose_fetcher',
```
Comment (msrb) [inline setup.py]: Hmm, can't help myself, but Python package names containing underscores look really weird to me :) I remember reading about some problems when some tools automatically replace underscores with dashes... just saying :)

File: hack/run_firehose_fetcher.sh [hunk 18]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+set -ex
+# As in worker, we need inject environment setup
+DIR=$(dirname "${BASH_SOURCE[0]}")
+source $DIR/env.sh
```
Comment (msrb) [inline hack/run_firehose_fetcher.sh]: Am I blind or did you forget to commit this file? :)

File: Dockerfile [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+COPY ./ /tmp/f8a_firehose-fetcher/
+RUN pushd /tmp/f8a_firehose-fetcher/ &&\
+  pip3 install . &&\
+  pip3 install --upgrade --no-binary :all: protobuf && pip3 install git+https://github.com/fabric8-analytics/fabric8-analytics-worker.git@${F8A_WORKER_VERSION} &&\
+  popd &&\
+  rm -rf /tmp/firehose_fetcher
+
+# A temporary hack to keep Selinon up2date
+COPY hack/update_selinon.sh /tmp/
```
Comment (msrb) [inline Dockerfile]: Do we need to install Selinon here? Any chance we could utilize docker cache by moving it before line 13?

File: Jenkinsfile [hunk 20]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env groovy
```
Comment (msrb) [inline Jenkinsfile]: I am afraid all new services will need to be compatible with CentOS CI. Which means no Jenkinsfile.

File: Jenkinsfile [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+        def dc = 'firehose-fetcher'
+        lock('f8a_staging') {
+
+            stage('Deploy - Stage') {
+                unstash 'template'
+                sh "oc --context=rh-idev process -v IMAGE_TAG=${commitId} -f template.yaml | oc --context=rh-idev apply -f -"
+            }
+
+            stage('End-to-End Tests') {
```
Comment (msrb) [inline Jenkinsfile]: Do you expect that there will be E2E tests covering this service? :)

File: openshift/template.yml (hunk 14 above)
Comment (fridex) [inline openshift/template.yml]: Yes, but they are needed in this container as well to correctly expand conf files (it will not affect functionality).

File: README.md [hunk 22]
```diff
@@ -0,0 +1,8 @@
+# Fabric8-Analytics Firehose Fetcher
+
+The aim of this service is to listen [Libraries.io Firehose Stream](https://github.com/librariesio/firehose-stream and based on events create new analyses.)
```
Comment (pkajaba) [inline README.md]: Good catch, thanks!

File: openshift/template.yml (hunk 7 above)
Comment (pkajaba) [inline openshift/template.yml]: Thanks, I based this template on jobs where it's bayesian-jobs. However, I will continue with f8a from now on.

https://github.com/fabric8-analytics/fabric8-analytics-jobs/blob/master/openshift/template.yaml

File: openshift/template.yml (hunk 8 above)
Comment (pkajaba) [inline openshift/template.yml]: not sure about this one @fridex can you tell?

File: openshift/template.yml (hunk 10 above)
Comment (pkajaba) [inline openshift/template.yml]: I recall that we need this one. @fridex what do you think?

File: Dockerfile (hunk 19 above)
Comment (pkajaba) [inline Dockerfile]: There is version defined in requirements.in of worker so it might override version which we installed in script.

**Comment by msrb** [comment]
LGTM, let's test this 😉
